### PR TITLE
[files_external] swift tests

### PR DESCRIPTION
**Pull request by MorrisJobke**
I just gave this a try, but I guess someone with more knowledge about Swift should have a look.

There seem to be some authentication problems. Is this url parameter the URL for getting the authentication? 

In the README of the docker container it is used like this:

```
vagrant@host1:~$ swift -A http://127.0.0.1:12345/auth/v1.0 -U test:tester -K testing stat
       Account: AUTH_test
    Containers: 0
       Objects: 0
         Bytes: 0
  Content-Type: text/plain; charset=utf-8
   X-Timestamp: 1402463864.77057
    X-Trans-Id: tx4e7861ebab8244c09dad9-005397e678
X-Put-Timestamp: 1402463864.77057
```

(-A to specify the URL for the authentication token retrieval, -U for user and -K for the key)

Is this then the correct syntax in our files_external code?

cc @butonic @DeepDiver1975

### CODE REVIEW

File: apps/files_external/tests/env/start-swift-morrisjobke.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+<?php
+
+return array(
+    'run'=>true,
+    'url'=>'http://$host:8080/auth/v1.0',
+    'user'=>'$user',
+    'key'=>'$password',
+    'bucket'=>'swift',
+    'region' => 'DFW',
+);
```
Comment (MorrisJobke) [inline apps/files_external/tests/env/start-swift-morrisjobke.sh]: THe above lines are the ones where I'm not sure if I did that correct.

Comment (MorrisJobke) [inline apps/files_external/tests/env/start-swift-morrisjobke.sh]: Maybe see https://github.com/ccollicutt/docker-swift-onlyone/issues/1

**Comment by PVince81** [comment]
What's the status ?

**Comment by MorrisJobke** [comment]
> What's the status ?

Needs further work, because I didn't figured out how to connect from ownCloud to swift ... maybe @Xenopathic can help here.

Maybe we can use the ceph/demo docker container that also exposes swift.

**Comment by Xenopathic** [comment]
So, it turns out the PHP library we use to connect to Swift only supports Keystone authentication (v2), not Swift built-in authentication (v1), which the current Docker image and Ceph support. It will be possible to get Keystone integrated with Ceph, but it could take some time...

**Comment by butonic** [comment]
@MorrisJobke nope. swift with ceph wont work because the libs we use don't support Keystone v1.

@Xenopathic I think time would be better invested in a devstack docker container. see https://wiki.openstack.org/wiki/Docker ... it seems several people are trying this: https://hub.docker.com/search/?q=devstack&page=1&isAutomated=0&isOfficial=0&starCount=0&pullCount=0

**Comment by Xenopathic** [comment]
@butonic With a bit of luck, setting up Keystone and connecting it to Ceph should be relatively straight-forward. Hopefully, I can avoid needing a MySQL database...: http://ceph.com/docs/master/radosgw/keystone/

**Comment by butonic** [comment]
@Xenopathic AFAIU you need to set up an openstack instance with keystone to make ceph hand over auth to that service. So ... twice the trouble. But pleaso prove me wrong ;)

**Comment by Xenopathic** [comment]
@butonic Sorry to break the bad news, but I just got Keystone working with Ceph without needing the whole OpenStack full-stack. I nearly swallowed my own tongue when the little green dot appeared...

**Comment by Xenopathic** [comment]
@MorrisJobke @butonic Get testing! The Docker image is available at https://hub.docker.com/r/xenopathic/ceph-keystone/, and I have full intentions of creating an S3 test around the same image.

**Comment by MorrisJobke** [comment]
> I nearly swallowed my own tongue when the little green dot appeared...

? Jenkins needs to be configured to actually run this one ;)

**Comment by Xenopathic** [comment]
@MorrisJobke No, in the admin interface, manually configuring it to see if it actually worked.

**Comment by Xenopathic** [comment]
@icewind1991 @MorrisJobke @butonic Review?

**Comment by icewind1991** [comment]
:+1: looks good from what I can tell

**Comment by PVince81** [comment]
Looks good, but if there are failing tests we can't merge it or put it into CI yet.

**Comment by DeepDiver1975** [comment]
there are failures:

https://ci.owncloud.org/job/server-master-linux-externals-swift/6/database=sqlite,external=swift-ceph,label=vm-slave-02/console

**Comment by DeepDiver1975** [comment]
@Xenopathic

**Comment by Xenopathic** [comment]
Hmm, I don't know what's changed, but I get an error while starting the Docker image:

> Unable to establish connection to http://172.17.0.2:35357/v2.0/OS-KSADM/services

Investigating... probably another timeout issue

**Comment by Xenopathic** [comment]
OK, those errors are now fixed (the previous sleep in the Docker image has been replaced with a netcat loop to check if the port is ready yet). Now to fix the *real* Swift errors:

```
There was 1 error:

1) Test\Files\Storage\Swift::testInstanceOfStorage
Guzzle\Http\Exception\ClientErrorResponseException: Client error response
[status code] 404
[reason phrase] Not Found
[url] http://172.17.0.7/swift/v1/swift

/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Exception/BadResponseException.php:43
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Message/Request.php:145
/home/users/robin/git/owncloud/core/3rdparty/symfony/event-dispatcher/Symfony/Component/EventDispatcher/EventDispatcher.php:164
/home/users/robin/git/owncloud/core/3rdparty/symfony/event-dispatcher/Symfony/Component/EventDispatcher/EventDispatcher.php:53
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Message/Request.php:589
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Message/Request.php:378
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Curl/CurlMulti.php:286
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Curl/CurlMulti.php:244
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Curl/CurlMulti.php:227
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Curl/CurlMulti.php:211
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Curl/CurlMulti.php:105
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Curl/CurlMultiProxy.php:91
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Client.php:282
/home/users/robin/git/owncloud/core/3rdparty/guzzle/http/Guzzle/Http/Message/Request.php:198
/home/users/robin/git/owncloud/core/3rdparty/rackspace/php-opencloud/lib/OpenCloud/ObjectStore/Resource/Container.php:274
/home/users/robin/git/owncloud/core/3rdparty/rackspace/php-opencloud/lib/OpenCloud/Common/Base.php:279
/home/users/robin/git/owncloud/core/3rdparty/rackspace/php-opencloud/lib/OpenCloud/ObjectStore/Resource/AbstractContainer.php:42
/home/users/robin/git/owncloud/core/3rdparty/rackspace/php-opencloud/lib/OpenCloud/ObjectStore/Resource/Container.php:50
/home/users/robin/git/owncloud/core/3rdparty/rackspace/php-opencloud/lib/OpenCloud/ObjectStore/Service.php:66
/home/users/robin/git/owncloud/core/apps/files_external/tests/backends/swift.php:45

--

There was 1 failure:

1) Test\Files\Storage\Swift::testStat
Failed asserting that false is true.

/home/users/robin/git/owncloud/core/tests/lib/files/storage/storage.php:311
```

**Comment by Xenopathic** [comment]
So I got annoyed that Docker would sometimes take longer to initialize, and fixed it once and for all. The start files use netcat (either traditional GNU or OpenBSD variants are supported) to wait for the port to be available before continuing. The logic works locally. Also, the Windows SMB tests will bail out gracefully if the server is unreachable, making a full test run much faster locally.

cc @MorrisJobke you'll probably like this

**Comment by Xenopathic** [comment]
Still one remaining Swift error though:

```
There was 1 failure:

1) Test\Files\Storage\Swift::testStat
Failed asserting that false is true.

/home/users/robin/git/owncloud/core/tests/lib/files/storage/storage.php:311
```

The reason for it is that Swift compares the filecache to a directory listing to check if anything has changed in hasUpdated() ( :see_no_evil: :hear_no_evil: ), which doesn't work in the unit tests (quite rightly too, the filecache has nothing to do with the storage implementation). *sigh*

**Comment by MorrisJobke** [comment]
> The reason for it is that Swift compares the filecache to a directory listing to check if anything has changed in hasUpdated() ( :see_no_evil: :hear_no_evil: ), which doesn't work in the unit tests (quite rightly too, the filecache has nothing to do with the storage implementation). sigh

@butonic @icewind1991 Can I ask you to comment on this?

**Comment by icewind1991** [comment]
You will have to overwrite the unit test to make it work

**Comment by Xenopathic** [comment]
@icewind1991 Looks like that's the only way we can proceed. Swift doesn't update directory mtimes at all (not even the immediate parent of files that are added/removed!). I did find an interesting issue report from 2014 however: https://github.com/owncloud/core/issues/8633. Basically, to fix this properly, we need a daemon running that listens for changes and can propagate them into ownCloud, #19166

**Comment by MorrisJobke** [comment]
For me the docker startup times out after 60 seconds :(

This is the keystone container log and it terminates after 45 seconds.
```
$ docker logs -f 443 # keystone container
creating /etc/ceph/ceph.client.admin.keyring
creating /etc/ceph/ceph.mon.keyring
monmaptool: monmap file /etc/ceph/monmap
monmaptool: set fsid to 44a837df-fa97-4793-b39c-4c8e9af0bbeb
monmaptool: writing epoch 0 to /etc/ceph/monmap (1 monitors)
creating /tmp/ceph.mon.keyring
importing contents of /etc/ceph/ceph.client.admin.keyring into /tmp/ceph.mon.keyring
importing contents of /etc/ceph/ceph.mon.keyring into /tmp/ceph.mon.keyring
ceph-mon: set fsid to 44a837df-fa97-4793-b39c-4c8e9af0bbeb
ceph-mon: created monfs at /var/lib/ceph/mon/ceph-443e1b24a3f1 for mon.443e1b24a3f1
set pool 0 size to 1
0
2015-09-24 14:40:39.612324 7f1d3d053900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-24 14:40:47.984836 7f1d3d053900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-24 14:40:48.012929 7f1d3d053900 -1 filestore(/var/lib/ceph/osd/ceph-0) could not find 23c2fcde/osd_superblock/0//-1 in index: (2) No such file or directory
2015-09-24 14:40:50.734659 7f1d3d053900 -1 created object store /var/lib/ceph/osd/ceph-0 journal /var/lib/ceph/osd/ceph-0/journal for osd.0 fsid 44a837df-fa97-4793-b39c-4c8e9af0bbeb
add item id 0 name 'osd.0' weight 1 at location {host=443e1b24a3f1,root=default} to crush map
starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /var/lib/ceph/osd/ceph-0/journal
starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /var/lib/ceph/osd/ceph-0/journal
pool 'cephfs_data' created
pool 'cephfs_metadata' created
new fs with metadata pool 2 and data pool 1
2015-09-24 14:40:59.112536 7f9365540800 -1 deprecation warning: MDS id 'mds.0' is invalid and will be forbidden in a future version.  MDS names may not start with a numeric digit.
starting mds.0 at :/0
Unable to establish connection to http://172.17.0.5:35357/v2.0/OS-KSADM/services
```

**Comment by butonic** [comment]
@MorrisJobke @Xenopathic same here:
```
                                                                                                                               ✚ ✱ ◼
creating /etc/ceph/ceph.client.admin.keyring
creating /etc/ceph/ceph.mon.keyring
monmaptool: monmap file /etc/ceph/monmap
monmaptool: set fsid to c8ccf898-c80b-4e65-a925-5224e4eb9e8f
monmaptool: writing epoch 0 to /etc/ceph/monmap (1 monitors)
creating /tmp/ceph.mon.keyring
importing contents of /etc/ceph/ceph.client.admin.keyring into /tmp/ceph.mon.keyring
importing contents of /etc/ceph/ceph.mon.keyring into /tmp/ceph.mon.keyring
ceph-mon: set fsid to c8ccf898-c80b-4e65-a925-5224e4eb9e8f
ceph-mon: created monfs at /var/lib/ceph/mon/ceph-6b94ab5ad780 for mon.6b94ab5ad780
set pool 0 size to 1
0
2015-09-24 14:46:26.464911 7fce18b46900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-24 14:46:26.716217 7fce18b46900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-09-24 14:46:26.720886 7fce18b46900 -1 filestore(/var/lib/ceph/osd/ceph-0) could not find 23c2fcde/osd_superblock/0//-1 in index: (2) No such file or directory
2015-09-24 14:46:26.937573 7fce18b46900 -1 created object store /var/lib/ceph/osd/ceph-0 journal /var/lib/ceph/osd/ceph-0/journal for osd.0 fsid c8ccf898-c80b-4e65-a925-5224e4eb9e8f
add item id 0 name 'osd.0' weight 1 at location {host=6b94ab5ad780,root=default} to crush map
starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /var/lib/ceph/osd/ceph-0/journal
starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /var/lib/ceph/osd/ceph-0/journal
pool 'cephfs_data' created
pool 'cephfs_metadata' created
new fs with metadata pool 2 and data pool 1
starting mds.0 at :/0
2015-09-24 14:46:31.724509 7f20e437a800 -1 deprecation warning: MDS id 'mds.0' is invalid and will be forbidden in a future version.  MDS names may not start with a numeric digit.
Unable to establish connection to http://172.17.0.7:35357/v2.0/OS-KSADM/services
```

**Comment by DeepDiver1975** [comment]
so let's rework on this in the scope of 9.0 - please focus on sev1 and sev2 issues - THX

**Comment by Xenopathic** [comment]
@MorrisJobke @butonic Cannot reproduce. Can you make sure you have the latest version of the docker image on your machine? Perhaps you still have a slightly older one which suffered from that problem, the new one has a netcat wait so it can never happen.

**Comment by butonic** [comment]
@MorrisJobke @Xenopathic the keystone-all process takes too long to start. raising the sleep 2 to sleep 10 helps.

**Comment by Xenopathic** [comment]
Ah, I wasn't running the right Docker commands to build and push the modified container that uses netcat to wait for the port. Should be fixed now!

**Comment by DeepDiver1975** [comment]
> Ah, I wasn't running the right Docker commands to build and push the modified container that uses netcat to wait for the port. Should be fixed now!

@Xenopathic I guess I was clear to not take care about this right now. Please focus on tasks related to 8.2 - THX

**Comment by Xenopathic** [comment]
@DeepDiver1975 I literally just forgot to push the Docker image I made a couple of days ago (https://github.com/owncloud/core/pull/14077#issuecomment-142702774), that was it.

**Comment by MorrisJobke** [comment]
Can we reuse the objectstore primary test setup here somehow? cc @Xenopathic @butonic

**Comment by Xenopathic** [comment]
@MorrisJobke The start/stop scripts are ever so slightly different, it might not be worth trying to combine them. These files are unlikely to change much, and when they do we can just port changes between them.

BTW, I've added you as a collaborator on the xenopathic/ceph-keystone image, so if any changes need to be made you can make them directly :smile:

**Comment by Xenopathic** [comment]
Can we get this retested and merged?

**Comment by MorrisJobke** [comment]
Docker logs:

```
docker logs -f aac
creating /etc/ceph/ceph.client.admin.keyring
creating /etc/ceph/ceph.mon.keyring
monmaptool: monmap file /etc/ceph/monmap
monmaptool: set fsid to 8ab4d030-204c-4627-b073-3c9bf6fd39e5
monmaptool: writing epoch 0 to /etc/ceph/monmap (1 monitors)
creating /tmp/ceph.mon.keyring
importing contents of /etc/ceph/ceph.client.admin.keyring into /tmp/ceph.mon.keyring
importing contents of /etc/ceph/ceph.mon.keyring into /tmp/ceph.mon.keyring
ceph-mon: set fsid to 8ab4d030-204c-4627-b073-3c9bf6fd39e5
ceph-mon: created monfs at /var/lib/ceph/mon/ceph-aacb8faf079a for mon.aacb8faf079a
set pool 0 size to 1
0
2015-11-04 15:39:54.405959 7f6c59cf2900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway




2015-11-04 15:39:56.372399 7f6c59cf2900 -1 journal FileJournal::_open: disabling aio for non-block journal.  Use journal_force_aio to force use of aio anyway
2015-11-04 15:39:56.399258 7f6c59cf2900 -1 filestore(/var/lib/ceph/osd/ceph-0) could not find 23c2fcde/osd_superblock/0//-1 in index: (2) No such file or directory
2015-11-04 15:39:57.550100 7f6c59cf2900 -1 created object store /var/lib/ceph/osd/ceph-0 journal /var/lib/ceph/osd/ceph-0/journal for osd.0 fsid 8ab4d030-204c-4627-b073-3c9bf6fd39e5
add item id 0 name 'osd.0' weight 1 at location {host=aacb8faf079a,root=default} to crush map
starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /var/lib/ceph/osd/ceph-0/journal
starting osd.0 at :/0 osd_data /var/lib/ceph/osd/ceph-0 /var/lib/ceph/osd/ceph-0/journal
pool 'cephfs_data' created
pool 'cephfs_metadata' created
new fs with metadata pool 2 and data pool 1
2015-11-04 15:40:04.905476 7fa7e97cf800 -1 deprecation warning: MDS id 'mds.0' is invalid and will be forbidden in a future version.  MDS names may not start with a numeric digit.
starting mds.0 at :/0
```

autotest logs:

```
Status: Image is up to date for xenopathic/ceph-keystone:latest
xenopathic/ceph-keystone container: aacb8faf079a91a7d793cf304777a16009d072a24481b905ed8100dc4751084e
Waiting for ceph initialization............................................................
[ERROR] Waited 60 seconds, no response
stop: stop-swift-ceph.sh
Docker executable found - stop and remove docker containers
Stopping and removing docker container aacb8faf079a91a7d793cf304777a16009d072a24481b905ed8100dc4751084e
Error response from daemon: Cannot destroy container aacb8faf079a91a7d793cf304777a16009d072a24481b905ed8100dc4751084e: Driver btrfs failed to remove root filesystem aacb8faf079a91a7d793cf304777a16009d072a24481b905ed8100dc4751084e: Failed to destroy btrfs snapshot: directory not empty
Error: failed to remove containers: [aacb8faf079a91a7d793cf304777a16009d072a24481b905ed8100dc4751084e]
rm: das Entfernen von „./../apps/files_external/tests//config.swift.php“ ist nicht möglich: Datei oder Verzeichnis nicht gefunden
skip: start-webdav-ownCloud.sh
```

**Comment by Xenopathic** [comment]
Is this that btrfs loopback thing with Ceph again?

**Comment by MorrisJobke** [comment]
> Is this that btrfs loopback thing with Ceph again?

Could be.

**Comment by PVince81** [comment]
Almost there:
```
Configuration read from /srv/www/htdocs/owncloud/tests/phpunit-autotest-external.xml

S................................................................ 65 / 73 ( 89%)
........

Time: 33.12 seconds, Memory: 24.00Mb

There was 1 skipped test:

1) Test\Files\Storage\Swift::testStat
Swift doesn't update the parents folder mtime

/srv/www/htdocs/owncloud/apps/files_external/tests/backends/swift.php:64

OK, but incomplete, skipped, or risky tests!
Tests: 73, Assertions: 323, Skipped: 1.
stop: stop-swift-ceph.sh
Docker executable found - stop and remove docker containers
Stopping and removing docker container 2659814329cba585215a939126e57e871b967ba95387d6bda99496ec86e35882
Error response from daemon: Cannot destroy container 2659814329cba585215a939126e57e871b967ba95387d6bda99496ec86e35882: Driver btrfs failed to remove root filesystem 2659814329cba585215a939126e57e871b967ba95387d6bda99496ec86e35882: Failed to destroy btrfs snapshot: directory not empty
Error: failed to remove containers: [2659814329cba585215a939126e57e871b967ba95387d6bda99496ec86e35882]
skip: start-webdav-ownCloud.sh
```

Tests passed, there was just an issue when cleaning up. Might not be critical.
This is with a BTRFS file system on the host.

**Comment by PVince81** [comment]
The error code is still 0, so if this is run on CI it would like interpret it as a success.

- [x] TEST: ./autotest-external.sh sqlite swift-ceph
- [x] TEST: ./autotest-external.sh sqlite amazons3-ceph

:+1: from me

**Comment by icewind1991** [comment]
:+1: looks good

**Comment by Xenopathic** [comment]
@MorrisJobke Please confirm if we need to port the Btrfs thing into this branch?

**Comment by MorrisJobke** [comment]
> @MorrisJobke Please confirm if we need to port the Btrfs thing into this branch?

It doesn't work here directly but I need to investigate further because the startup of the container looks good (It needed a longer time to start than 60 seconds).

I would accept this, if this runs fine on Jenkins. @DeepDiver1975 Can you setup the CI run for this? Thanks

**Comment by PVince81** [comment]
Can we move this forward ? There are at least 4 improvement PRs for SWIFT that are waiting for these tests.

**Comment by MorrisJobke** [comment]
I'm fine with merging. @DeepDiver1975 The CI job shouldn't be that hard to setup, right? :wink:

**Comment by DeepDiver1975** [comment]
> I'm fine with merging. @DeepDiver1975 The CI job shouldn't be that hard to setup, right? :wink:

no

**Comment by MorrisJobke** [comment]
> no

Then please merge ;)

**Pull request merged**